xds/internal/balancer/gracefulswitch/gracefulswitch_test.go

```diff
@@ -0,0 +1,1354 @@
+	return balancer.PickResult{}, nil
+}
+
+// TestUpdateSubConnStateRace tests the race condition when the GracefulSwitchBalancer
+// receives a SubConnUpdate concurrently with an UpdateState() call, which can cause
+// the balancer to forward the update to to be closed and cleared. The balancer API
+// guarantees to never call any method the balancer after a Close() call, and the test
+// verifies that doesn't happen within the Graceful Switch Load Balancer.
+func (s) TestUpdateSubConnStateRace(t *testing.T) {
+	tcc, gsb := setup(t)
+	builder := balancer.Get(verifyBalName)
+	gsb.SwitchTo(builder)
+
+	builder = balancer.Get(balancerName1)
+	gsb.SwitchTo(builder)
+	currBal := gsb.balancerCurrent.Balancer.(*verifyBalancer)
+	currBal.t = t
+	pendBal := gsb.balancerPending.Balancer.(*mockBalancer1)
+	sc, err := currBal.newSubConn([]resolver.Address{}, balancer.NewSubConnOptions{})
+	if err != nil {
+		t.Fatalf("error constructing newSubConn in gsb: %v", err)
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	select {
+	case <-ctx.Done():
+		t.Fatalf("timeout while waiting for an NewSubConn call on the ClientConn")
+	case <-tcc.NewSubConnCh:
+	}
+	// Spawn a goroutine that constantly calls UpdateSubConn for the current
+	// balancer, which will get deleted in this testing goroutine.
+	finished := make(chan struct{})
+	go func() {
+		for {
+			select {
+			case <-finished:
+				return
+			default:
+			}
+			gsb.UpdateSubConnState(sc, balancer.SubConnState{
+				ConnectivityState: connectivity.Ready,
+			})
+		}
+	}()
+	time.Sleep(time.Millisecond)
+	// This UpdateState call causes current to be closed/cleared.
+	pendBal.updateState(balancer.State{
+		ConnectivityState: connectivity.Ready,
+	})
+	// From this, either one of two things happen. Either the
+	// GracefulSwitchBalancer doesn't Close() the current balancer before it
+	// forwards the SubConn update to the child, and the call gets forwarded
+	// down to the current balancer, or it can Close() the current balancer in
+	// between reading the balancer pointer and writing to it, and in that case
+	// the old current balancer should not be updated, as the balancer has
+	// already been closed and the balancer API guarantees it.
+	close(finished)
+}
+
+// TestInlineCallbackInBuild tests the scenario where a balancer calls back into
+// the balancer.ClientConn API inline from it's build function.
+func (s) TestInlineCallbackInBuild(t *testing.T) {
+	tcc, gsb := setup(t)
+	builder := balancer.Get(buildCallbackBalName)
+	// This build call should cause all of the inline updates to forward to the
+	// ClientConn.
+	gsb.SwitchTo(builder)
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	select {
+	case <-ctx.Done():
+		t.Fatalf("timeout while waiting for an UpdateState() call on the ClientConn")
+	case <-tcc.NewStateCh:
+	}
+	ctx, cancel = context.WithTimeout(context.Background(), defaultTestTimeout)
```

Comment:
No need to recreate the context.  Keep using the existing one.  It can be used for the whole test.